From db317f7119019d6274fa29070efb43424144b1b2 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Benno=20F=C3=BCnfst=C3=BCck?= <benno.fuenfstueck@neodyme.io>
Date: Tue, 14 Apr 2026 12:29:30 +0200
Subject: [PATCH] renderer: fix cases of wrong cursor orientation

If Y_0_TOP is not set, we should not flip the cursor data. Downstream consumers like QEMU expect the returned buffer to be the cursor in a format where y=0 is bottom. This fixes a bug where QEMU renders an inverted mouse cursor if using a gl-accelerated display with certain wayland compositors (for example wl-roots based) in the guest.

Part-of: <https://gitlab.freedesktop.org/virgl/virglrenderer/-/merge_requests/1611>
---
 src/vrend/vrend_renderer.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/src/vrend/vrend_renderer.c b/src/vrend/vrend_renderer.c
index 9938fa74..562efdc1 100644
--- a/src/vrend/vrend_renderer.c
+++ b/src/vrend/vrend_renderer.c
@@ -13023,7 +13023,8 @@ void *vrend_renderer_get_cursor_contents(struct pipe_resource *pres,
    }
 
    for (h = 0; h < res->base.height0; h++) {
-      uint32_t doff = (res->base.height0 - h - 1) * res->base.width0 * blsize;
+      uint32_t dh = res->y_0_top ? (res->base.height0 - h - 1) : (h);
+      uint32_t doff = dh * res->base.width0 * blsize;
       uint32_t soff = h * res->base.width0 * blsize;
 
       memcpy(data2 + doff, data + soff, res->base.width0 * blsize);
-- 
2.55.0

